core/src/actors/chain_manager/handlers.rs

```diff
@@ -71,37 +73,62 @@ impl Handler<GetHighestCheckpointBeacon> for ChainManager {
 
 /// Handler for AddNewBlock message
 impl Handler<AddNewBlock> for ChainManager {
-    type Result = Result<Hash, ChainManagerError>;
-
-    fn handle(
-        &mut self,
-        msg: AddNewBlock,
-        _ctx: &mut Context<Self>,
-    ) -> Result<Hash, ChainManagerError> {
-        let res = self.process_new_block(msg.block);
-        match res {
-            Ok(hash) => {
-                // Get SessionsManager's address
-                let sessions_manager_addr = System::current().registry().get::<SessionsManager>();
-
-                // Tell SessionsManager to announce the new block through every consolidated Session
-                let items = vec![InventoryEntry::Block(hash)];
-                sessions_manager_addr.do_send(Broadcast {
-                    command: AnnounceItems { items },
-                });
-            }
-            Err(ChainManagerError::BlockAlreadyExists) => {
-                debug!("Block already exists");
-            }
-            Err(ChainManagerError::StorageError(_)) => {
-                debug!("Error when serializing block");
-            }
-            Err(_) => {
-                debug!("Unexpected error");
-            }
-        };
-
-        res
+    type Result = SessionUnitResult;
+
+    fn handle(&mut self, msg: AddNewBlock, ctx: &mut Context<Self>) {
+        // Block verify process
+        let reputation_manager_addr = System::current().registry().get::<ReputationManager>();
+
+        reputation_manager_addr
+            .send(ValidatePoE {
+                block_header: msg.block.header.clone(),
+            })
+            .into_actor(self)
+            .then(|res, act, _ctx| {
+                match res {
+                    Err(e) => {
+                        // Error when sending message
+                        error!("Unsuccessful communication with reputation manager: {}", e);
+                    }
+                    Ok(valid_poe) => {
+                        let block_epoch = msg.block.header.block_header.beacon.checkpoint;
+                        if block_epoch == act.current_epoch {
```

Comment:
This verification should be made before querying `ReputationManager` at all in order to save unnecessary asynchronous IO.